### ISSUE TITLE
Error on plugin activation

### ISSUE THREAD

**Issue by moussetc** [mattermost/mattermost-plugin-sample#17]
When trying out a self-compiled version of the master branch of the sample plugin, there are the following errors:
- When enabling the plugin:
```
{"level":"error","ts":1541187867.5791962,"caller":"plugin/client_rpc_generated.go:67","msg":"RPC call OnConfigurationChange to plugin failed.","plugin_id":"com.mattermost.sample-plugin","error":"unexpected EOF"}
{"level":"error","ts":1541187868.0012312,"caller":"plugin/client_rpc_generated.go:67","msg":"RPC call OnConfigurationChange to plugin failed.","plugin_id":"com.mattermost.sample-plugin","error":"connection is shut down"}
{"level":"error","ts":1541187868.238017,"caller":"plugin/client_rpc_generated.go:67","msg":"RPC call OnConfigurationChange to plugin failed.","plugin_id":"com.mattermost.sample-plugin","error":"connection is shut down"}
```

- When disabling the plugin:
```
{"level":"warn","ts":1541187955.0697181,"caller":"go-plugin/client.go:371","msg":"error closing client during Kill","plugin_id":"com.mattermost.sample-plugin","wrapped_extras":"errconnection is shut down"}`
```

Mattermost Server is a 5.4.0-rc1 Docker running on a Manjaro.
Does anyone else have that issue?

**Comment by hanzei** [comment]
Some one the forum encountered the same problem: https://forum.mattermost.org/t/sample-plugin-problem/6054.

**Comment by hanzei** [comment]
Would you please post the sections about plugins of your Mattermost  `config.json`? This might help debugging this.

**Comment by moussetc** [comment]
Here's the plugins configuration:
``` 
"PluginSettings": {
        "Enable": true,
        "EnableUploads": true,
        "Directory": "./plugins",
        "ClientDirectory": "./client/plugins",
        "Plugins": {},
        "PluginStates": {
            "com.mattermost.sample-plugin": {
                "Enable": true
            }
        }
    },
```

**Comment by hanzei** [comment]
I was able to reproduce the error with the latest Mattermost master. I also found some interesting log messages:
```
2018-11-08T12:38:38.976+0100	debug	go-plugin/client.go:823	panic: setConfiguration called with the existing configuration	{"plugin_id": "com.mattermost.sample-plugin"}
2018-11-08T12:38:38.976+0100	debug	go-plugin/client.go:823		{"plugin_id": "com.mattermost.sample-plugin"}
2018-11-08T12:38:38.976+0100	debug	go-plugin/client.go:823	goroutine 41 [running]:	{"plugin_id": "com.mattermost.sample-plugin"}
2018-11-08T12:38:38.976+0100	debug	go-plugin/client.go:823	main.(*Plugin).setConfiguration(0xc000131ac0, 0x10bcc48)	{"plugin_id": "com.mattermost.sample-plugin"}
2018-11-08T12:38:38.977+0100	debug	go-plugin/client.go:823		/path/to/go/src/github.com/mattermost/plugin-sample/server/configuration.go:56 +0xd3	{"plugin_id": "com.mattermost.sample-plugin"}
2018-11-08T12:38:38.977+0100	debug	go-plugin/client.go:823	main.(*Plugin).OnConfigurationChange(0xc000131ac0, 0xa86040, 0xc000131ac0)	{"plugin_id": "com.mattermost.sample-plugin"}
2018-11-08T12:38:38.977+0100	debug	go-plugin/client.go:823		/path/to/go/src/github.com/mattermost/plugin-sample/server/configuration.go:71 +0xcd	{"plugin_id": "com.mattermost.sample-plugin"}
2018-11-08T12:38:38.977+0100	debug	go-plugin/client.go:823	github.com/mattermost/plugin-sample/server/vendor/github.com/mattermost/mattermost-server/plugin.(*hooksRPCServer).OnConfigurationChange(0xc000148580, 0x10bcc48, 0xc00013e540, 0x0, 0x0)	{"plugin_id": "com.mattermost.sample-plugin"}
2018-11-08T12:38:38.977+0100	debug	go-plugin/client.go:823		/path/to/go/src/github.com/mattermost/plugin-sample/server/vendor/github.com/mattermost/mattermost-server/plugin/client_rpc_generated.go:77 +0x69	{"plugin_id": "com.mattermost.sample-plugin"}
2018-11-08T12:38:38.977+0100	debug	go-plugin/client.go:823	reflect.Value.call(0xc000072c00, 0xc00000e670, 0x13, 0xb006b2, 0x4, 0xc00023ef18, 0x3, 0x3, 0xc000144240, 0x0, ...)	{"plugin_id": "com.mattermost.sample-plugin"}
2018-11-08T12:38:38.977+0100	debug	go-plugin/client.go:823		/path/to/.local/go/src/reflect/value.go:447 +0x449	{"plugin_id": "com.mattermost.sample-plugin"}
2018-11-08T12:38:38.977+0100	debug	go-plugin/client.go:823	reflect.Value.Call(0xc000072c00, 0xc00000e670, 0x13, 0xc00026af18, 0x3, 0x3, 0x0, 0x0, 0x0)	{"plugin_id": "com.mattermost.sample-plugin"}
2018-11-08T12:38:38.977+0100	debug	go-plugin/client.go:823		/path/to/.local/go/src/reflect/value.go:308 +0xa4	{"plugin_id": "com.mattermost.sample-plugin"}
2018-11-08T12:38:38.977+0100	debug	go-plugin/client.go:823	net/rpc.(*service).call(0xc00006f6c0, 0xc0000a6640, 0xc000026940, 0xc000026950, 0xc000114500, 0xc00000d160, 0x9dad00, 0x10bcc48, 0x16, 0x9dad40, ...){"plugin_id": "com.mattermost.sample-plugin"}
2018-11-08T12:38:38.977+0100	debug	go-plugin/client.go:823		/path/to/.local/go/src/net/rpc/server.go:384 +0x14e	{"plugin_id": "com.mattermost.sample-plugin"}
2018-11-08T12:38:38.977+0100	debug	go-plugin/client.go:823	created by net/rpc.(*Server).ServeCodec	{"plugin_id": "com.mattermost.sample-plugin"}
2018-11-08T12:38:38.977+0100	debug	go-plugin/client.go:823		/path/to/.local/go/src/net/rpc/server.go:481 +0x47e	{"plugin_id": "com.mattermost.sample-plugin"}
2018-11-08T12:38:38.978+0100	error	plugin/client_rpc_generated.go:67	RPC call OnConfigurationChange to plugin failed.	{"plugin_id": "com.mattermost.sample-plugin", "error": "unexpected EOF"}
```
@moussetc Do you see similar messages in your log? They should appear above the ones you posted.

@lieut-data It appears to be [this](https://github.com/mattermost/mattermost-plugin-sample/blob/master/server/configuration.go#L56) loc which throws the panic. But I thought this should not happen. Do you have any idea why this happens?

**Comment by lieut-data** [comment]
Thanks for the heads up, @hanzei! Not sure if I just missed this when we ported the configuration changes to the sample plugin, or if it's a recent Go optimization, but the empty struct has zero width, allowing Go to allocate the same memory address each time `new(configuration)` is called.

I've submitted a [PR](https://github.com/mattermost/mattermost-plugin-sample/pull/21) to avoid this, since it's probably a common case for a plugin /not/ to have a configuration. We should probably backport this to the demo plugin too just in case someone starts from there.
